app/src/main/java/com/quran/labs/androidquran/widgets/HighlightingImageView.java

```diff
@@ -130,17 +152,178 @@ public void setNightMode(boolean isNightMode, int textBrightness) {
     adjustNightMode();
   }
 
-  public void highlightAyah(int sura, int ayah, HighlightType type) {
+  private void highlightFloatableAyah(Set<String> highlights, String currentSurahAyah) {
+    final Map<String, List<AyahBounds>> coordinatesData = ayahCoordinates.getAyahCoordinates();
+
+    String previousSurahAyah = currentSurahAyah;
+    for(String surahAyahs: highlights) {
+      previousSurahAyah = surahAyahs;
+    }
+    List<AyahBounds> startingBounds;
+    int arrowIndex = previousSurahAyah.indexOf("->");
+    if(arrowIndex != -1) {
+      // The ayah changed during animating
+      startingBounds = (List<AyahBounds>)animator.getAnimatedValue();
+      animator.cancel();
+      previousSurahAyah = previousSurahAyah.substring(arrowIndex + 2);
+    } else {
+      startingBounds = coordinatesData.get(previousSurahAyah);
+    }
+
+    final String ayahTransition = previousSurahAyah+"->"+currentSurahAyah;
+
+    // yes we make copies, because normalizing the bounds will change them
+    List<AyahBounds> previousAyahBoundsList = new ArrayList<>(startingBounds);
+    List<AyahBounds> currentAyahBoundsList = new ArrayList<>(coordinatesData.get(currentSurahAyah));
+
+    highlights.clear();
+    highlights.add(ayahTransition);
+
+    // TODO: reason whether this TypeEvaluator should be anonymous or not
+    animator = ValueAnimator.ofObject(new TypeEvaluator() {
+
+      private void normalizeAyahBoundsList(List<AyahBounds> start, List<AyahBounds> end) {
+        // this function takes two unequal length lists and normalizes them
+
+        int startSize = start.size();
+        int endSize = end.size();
+        int minSize = Math.min(startSize, endSize);
+        int maxSize = Math.max(startSize, endSize);
+        List<AyahBounds> minList = startSize < endSize? start : end;
+        int diff = maxSize - minSize;
+
+        RectF rectToBeDivided = minList.get(minSize-1).getBounds();
+        float oLeft = rectToBeDivided.left;
+        float oRight = rectToBeDivided.right;
+        float oTop = rectToBeDivided.top;
+        float oBottom = rectToBeDivided.bottom;
+        minList.remove(minSize-1);
+        float part = (oRight-oLeft) /(diff+1);
+        for(int i=0; i<(diff+1); ++i) {
+          float left = oLeft + part*i;
+          float right = left + part;
+          RectF rect = new RectF(left, oTop, right, oBottom);
+          AyahBounds ayahBounds = new AyahBounds(0, 0, rect);
+          minList.add(ayahBounds);
+        }
+      }
+
+      @Override
+      public Object evaluate(float fraction, Object startObject, Object endObject) {
+        List<AyahBounds> start = (List<AyahBounds>)startObject;
+        List<AyahBounds> end = (List<AyahBounds>)endObject;
+
+        if(start.size() != end.size()) {
+          normalizeAyahBoundsList(start, end);
+        }
+
+        int size = start.size();
+
+        //TODO: instead of creating result every time, modify in-place
```

Comment:
this matters too because we're in `onDraw` here potentially

Reply:
Well, the `evaluate()` function here, provides new values. The `onAnimationUpdate()` hook lets the user use those values and it runs in the UI thread.

I have noticed during my debugging sessions that the `evaluate()` gets called more times than `onAnimationUpdate()` therefore, they are potentially running independent of each other, which makes sense because `onAnimationUpdate()` will be called in an optimized way (once per render frame) but `evaluate()` can be called every update frame. So, we need not worry about our draws getting janked due to `evaluate()` as they are independent of each other.

Moreover, even if we ignore the independence, we can't ignore the fact that if we modify the result in place then we will have a data race condition between the writer (`evaluate()`) and the reader (`onAnimationUpdate()`). We might have a rare  case in which the `result` is simultaneously accessed by `onAnimationUpdate()` and also written by `evaluate()`

So, I have simply removed the TODO and haven't changed the code.